### PR TITLE
Extract Worker actor from Staker

### LINKED ISSUE #800: Find a better metaphor than "miner"
I think this has been eloquently described by @jMyles:

> (...) it is more clear to me than ever that `Miner` is a broken metaphor.  I have never liked this name, but reading this, I am given pause that there is a number mismatch between the work types for a given node and the practices surrounding this work types.  For example, the inflation economics apply to a "Staker" or "Worker" (half of what we currently call `Miner`).  Since this is a work token, `Worker` makes sense to me.  However, the _work_ being performed pursuant to this work token, as it relates to the adjudication, is not of only one kind.  Presently, we expect `Ursula` to perform PRE and understand that she'll be slashed if she fails to perform.  Let's imagine a future in which we task a different character to perform SSS or NuFHE under the auspices of the same work token.  Is this person still a `Miner`?  It seems to me that `Worker` is a concept that might _utilize_ `Ursulas` and SSS/NuFHE-characters and is not properly understood as a "mining mixin."  Normally, I am not terribly concerned about these sorts of future-proofing exercises, but I think that we'll have crossed an ~epitaph~ rubicon if we deploy these smart contracts with the word "Miner" - it will be hard to educate our audience on what we really mean.

### LINKED ISSUE #962: How should Bob discover and validate Ursula's Identity Evidence?
_'Identity evidence'_ (sometimes called _'Evidence of decentralized identity'_) is a signature of Ursula's stamp with the checksum address.

Perhaps I'm too slow today, but I can't figure out what's the best way for Bob to discover this information in a way that gives him assurance that it's valid and compatible with the information contained in `MinersEscrow` contract.

Bob needs this information in order to request a CFrag evaluation and Ursula's slashing.

### LINKED ISSUE #1003: Re-design Ursula's identity evidence
We have to come up with a new design of identity evidence that fulfills with, at least, the following requirements:
* Allows Bob (and other characters) to check that the Ursula node they want to validate belongs to a valid staker. This is currently not checked anywhere.
* Allows stakers to be smart contracts, which can't sign stuff. This is what #949 does, but we will probably need to rethink that approach.
----
**Current approach (also followed by #996):**
* Ursulas have a non-ETH signing key (_the stamp_), which is signed with the ETH key of the staker to produce her _identity evidence_. The stamp is used for signing network protocol messages. 
* Ursulas inherit the address associated to the staker's ETH key (_the checksum address_) and use it as their ID. Ursulas don't need to control the staker's ETH key.
* When Bob validates an Ursula node, he only checks that the signature is valid, but he doesn't check that the staker's account is valid (e.g., it can be a signature from an arbitrary ETH account, with no tokens at all).

**#949's approach:**
* Ursulas have a stamp and an ETH key (which results in a worker address)
* Stakers have an ETH account (with or without keys, since they can be contracts).
* `MinersEscrow` has a list of the mapping between Ursulas' addresses and stakers' addresses.
* This new scheme is not connected to the network-level identity information, so it's not consistent at all. The fact that this doesn't raise any problems is bad, since it shows that we weren't really validating anything with our current identity evidence approach.
-----
**New approach?**
We can mix both approaches in something like this:
* Ursula does not need an ETH address, only her stamp, like it is now.
* `MinersEscrow` has a list of the mapping between Ursulas' stamps and stakers addresses.
* Bob, Alice and other characters should always check with `MinersEscrow` that a node they have learnt about is actually connected to a staking miner. This means characters need access to an ETH node for this validation.
* The ETH signature as proof of identity is not useful anymore, as even if it's valid, you have to check `MinersEscrow`.
* Ursula can continue using the staker's address as her `checksum_public_address`, even if it's a contract.

### LINKED ISSUE #1033: Check valid staker->worker relation in Learner._is_valid_worker() against StakingEscrow
This can be done by checking `StakingEscrow.workerToStaker(worker_address) == staker_address` or `StakingEscrow.stakerInfo[staker_address].worker == worker_address`.

The former seems cheaper.

### LINKED ISSUE #1054: Extract Worker actor from current Staker
With the introduction of a Worker role in our smart contracts as part of the Alhambra Verde EPIC, we need to separate our `Staker` actor class in two parts: The real `Staker`, and the `Worker`. 

For starters, most (all?) methods in `Staker` marked with the `@only_me` decorator seems like  suitable candidates to remain in `Staker`. However, it seems reasonable to keep some of the "read-only" methods in `Staker` for `Worker` (i.e., `get_locked_tokens`).

Perhaps a "stranger" Staker (i.e., `is_me=False`) can be part of a `Worker`?

### PR DESCRIPTION
**Pull request by cygnusv**
The goal of this PR is to crystallize the implications of #1003, i.e., the separation of Ursula in two: Worker and Staker. Now Ursulas are worker nodes, while NU holders are stakers (either of human or contract nature).

Most drastic changes occur at the actor level (since now the old `Staker` actor was separated into `Staker` and `Worker`) and at the `Teacher` level, since now teacher nodes (which inherit from `Worker`) have to implement the new ways of identity validation (as per #1003), which include off-chain and on-chain means.

Side note: this was a fun PR. @KPrasch and I were working on it in turns, taking advantage of our almost complementary time zones. There wasn't too much need to talk. Just wake up in the morning, drink some coffee, `git remote update` and continue from where the other left. 

- Closes #1003 
- Closes #1033 
- Closes #1054
- Closes #962 
- Closes #800

### CODE REVIEW

File: nucypher/blockchain/eth/actors.py [hunk 1]
```diff
@@ -686,11 +645,155 @@ def collect_policy_reward(self, collector_address=None, policy_agent: PolicyAgen
...
         self._transaction_cache.append((datetime.utcnow(), collection_txhash))
         return collection_txhash
 
 
+class Worker(NucypherTokenActor):
+    """
+    Ursula baseclass for blockchain operations, practically carrying a pickaxe.
+    """
+
+    __current_period_sample_rate = 60*60  # seconds
```
Comment (KPrasch) [inline nucypher/blockchain/eth/actors.py]: Perhaps - Let's change this is be based on a `TokenEconomics` instance?

File: tests/characters/test_ursula_prepares_to_act_as_mining_node.py [hunk 2]
```diff
@@ -124,3 +137,34 @@ def test_vladimir_uses_his_own_signing_key(blockchain_alice, blockchain_ursulas)
     # However, the actual handshake proves him wrong.
     with pytest.raises(vladimir.InvalidNode):
         vladimir.verify_node(blockchain_alice.network_middleware, certificate_filepath="doesn't matter")
+
+
+# TODO: Change name of this file
+def test_blockchain_ursulas_reencrypt(blockchain_ursulas, blockchain_alice, blockchain_bob, policy_value):
```
Comment (KPrasch) [inline tests/characters/test_ursula_prepares_to_act_as_mining_node.py]: :boom:

**Comment by KPrasch** [comment]
Follow up to `@transaction` agency API - https://github.com/nucypher/nucypher/issues/1086

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol [hunk 3]
```diff
@@ -357,25 +357,33 @@ contract StakingEscrow is Issuer {
...
+            // Remove relation between the old worker and the staker, if there was one
+            if (info.worker != address(0)) {
+                workerToStaker[info.worker] = address(0);
+            }
+            info.worker = _worker;
+            info.workerStartPeriod = currentPeriod;
+            // Set new worker->staker relation
+            workerToStaker[_worker] = msg.sender;
+        } else { // Staker wants to unset her worker
+            // In this case, we don't require a minimum elapsed time with the worker
```
Comment (szotov) [inline nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol]: When staker changes worker - he potentially could avoid punishment. I think only setting from zero address to non-zero can be without checking period

**Pull request merged**